include/networkit/algebraic/CSRGeneralMatrix.hpp

```diff
@@ -0,0 +1,1216 @@
+/*
+ * CSRGeneralMatrix.hpp
+ *
+ *  Created on: May 6, 2015
+ *     Authors: Michael Wegner <michael.wegner@student.kit.edu>
+ *              Eugenio Angriman <angrimae@hu-berlin.de>
+ */
+
+#ifndef CSRGENERALMATRIX_H_
+#define CSRGENERALMATRIX_H_
+
+#include <omp.h>
+#include <vector>
+
+#include <networkit/Globals.hpp>
+#include <networkit/algebraic/AlgebraicGlobals.hpp>
+#include <networkit/algebraic/Vector.hpp>
+#include <networkit/graph/Graph.hpp>
+#include <networkit/algebraic/SparseAccumulator.hpp>
+#include <networkit/auxiliary/Timer.hpp>
+
+namespace NetworKit {
+
+/**
+ * @ingroup algebraic
+ * The CSRGeneralMatrix class represents a sparse matrix stored in CSR-Format (i.e. compressed sparse row).
+ * If speed is important, use this CSRGeneralMatrix instead of the Matrix class.
+ */
+template <typename ValueType>
+class CSRGeneralMatrix {
+private:
+	std::vector<index> rowIdx;
+	std::vector<index> columnIdx;
+	std::vector<ValueType> nonZeros;
+
+	count nRows;
+	count nCols;
+	bool isSorted;
+	ValueType zero;
+
+	/**
+	 * Quicksort algorithm on columnIdx between [@a left, @a right].
+	 * @param left
+	 * @param right
+	 */
+	void quicksort(index left, index right) {
+		if (left >= right) return;
+		index pivotIdx = partition(left, right);
+		if (pivotIdx != 0) {
+			quicksort(left, pivotIdx-1);
+		}
+		quicksort(pivotIdx+1, right);
+	}
+
+
+	/**
+	 * Partitions columnIdx between [@a left, @a right] after selecting the pivot in the middle.
+	 * @param left
+	 * @param right
+	 * @return The pivot.
+	 */
+	index partition(index left, index right) {
+		index mid = (left + right) / 2;
+		index pivot = columnIdx[mid];
+		std::swap(columnIdx[mid], columnIdx[right]);
+		std::swap(nonZeros[mid], nonZeros[right]);
+
+		index i = left;
+		for (index j = left; j < right; ++j) {
+			if (columnIdx[j] <= pivot) {
+				std::swap(columnIdx[i], columnIdx[j]);
+				std::swap(nonZeros[i], nonZeros[j]);
+				++i;
+			}
+		}
+		std::swap(columnIdx[i], columnIdx[right]);
+		std::swap(nonZeros[i], nonZeros[right]);
+		return i;
+		}
+
+
+	/**
+	 * Binary search the sorted columnIdx vector between [@a left, @a right] for column @a j.
+	 * If @a j is not present, the index that is immediately left of the place where @a j would be
+	 * is returned. If
+	 * @param left
+	 * @param right
+	 * @param j
+	 * @return The position of column @a j in columnIdx or the element immediately to the left of the place where @a j
+	 * would be.
+	 */
+	index binarySearchColumns(index left, index right, index j) const {
+	assert(sorted());
```

Comment:
Indentation is broken here (and in large parts of the remainder of the file). Please fix it and make it consistent.